aspnetcore/fundamentals/middleware/extensibility.md

```diff
@@ -0,0 +1,118 @@
+
+[View or download sample code](https://github.com/aspnet/Docs/tree/master/aspnetcore/fundamentals/middleware/extensibility/sample) ([how to download](xref:tutorials/index#how-to-download-a-sample))
+
+The sample app demonstrates:
+
+* Conventionally-activated middleware.
+* Middleware activated by an `IMiddlewareFactory` registered in an [Ninject](http://www.ninject.org/) container.
+
+The middlewares function identically and record the value provided by a query string parameter (`key`). The middlewares use an injected database context (a scoped service) to record the query string value in an in-memory database.
+
+## IMiddleware
+
+[IMiddleware](/dotnet/api/microsoft.aspnetcore.http.imiddleware) defines middleware for the app's request pipeline. The [InvokeAsync(HttpContext, RequestDelegate)](/dotnet/api/microsoft.aspnetcore.http.imiddleware.invokeasync#Microsoft_AspNetCore_Http_IMiddleware_InvokeAsync_Microsoft_AspNetCore_Http_HttpContext_Microsoft_AspNetCore_Http_RequestDelegate_) method handles requests and returns a `Task` that represents the execution of the middleware.
+
+Conventionally-activated middleware:
+
+[!code-csharp[Main](extensibility/sample/Middleware/MiddlewareViaConventionalActivation.cs?name=snippet1)]
+
+`IMiddlewareFactory`-activated middleware:
+
+[!code-csharp[Main](extensibility/sample/Middleware/MiddlewareViaIMiddlewareFactoryActivation.cs?name=snippet1)]
+
+Extensions are created for the middlewares:
+
+[!code-csharp[Main](extensibility/sample/Middleware/MiddlewareExtensions.cs?name=snippet1)]
+
+Note that it isn't possible to pass objects with `UseMiddleware` as it is with convention-based middleware activation:
+
+```csharp
+public static IApplicationBuilder UseMiddlewareViaIMiddlewareFactoryActivation(
+    this IApplicationBuilder builder, bool option)
+{
+    // Passing 'option' as an argument won't work and throws a NotSupportedException 
+    // at runtime.
+    return builder.UseMiddleware<MiddlewareViaIMiddlewareFactoryActivation>(option);
+}
+```
+
+The middlewares are added to the built-in container in *Startup.cs*:
+
+[!code-csharp[Main](extensibility/sample/Startup.cs?name=snippet1&highlight=6-7)]
+
+The middlewares are registered in the request processing pipeline in *Startup.cs*:
+
+[!code-csharp[Main](extensibility/sample/Startup.cs?name=snippet2&highlight=15-16)]
+
+## IMiddlewareFactory
+
+[IMiddlewareFactory](/dotnet/api/microsoft.aspnetcore.http.imiddlewarefactory) provides methods to create middleware. The middleware factory is registered in the nonconforming container as a scoped service.
+
+The default `IMiddlewareFactory` implementation is found in the [Microsoft.AspNetCore.Http](https://www.nuget.org/packages/Microsoft.AspNetCore.Http/) package:
+
+```csharp
+public class MiddlewareFactory : IMiddlewareFactory
```

Comment:
This is an implementation detail. I'm not sure we should have the code spelled out in the docs unless it's meant to be "similar" but not exact.

Reply:
Sure ... I'll cut this. It was just meant to aid the reader wishing to see it without forcing them to click into reference source.